lib/core/batchingReceiver.ts

```diff
@@ -79,35 +79,40 @@ export class BatchingReceiver extends MessageReceiver {
       let onReceiveMessage: OnAmqpEventAsPromise;
       let onSessionClose: OnAmqpEventAsPromise;
       let onReceiveClose: OnAmqpEventAsPromise;
+      let onReceiveDrain: OnAmqpEvent;
       let onReceiveError: OnAmqpEvent;
       let onSessionError: OnAmqpEvent;
       let waitTimer: any;
       let maxMessageWaitTimer: any;
-      const resetCreditWindow = () => {
-        this._receiver!.setCreditWindow(0);
-        this._receiver!.addCredit(0);
-      };
+
       // Final action to be performed after maxMessageCount is reached or the maxWaitTime is over.
       const finalAction = () => {
         if (maxMessageWaitTimer) {
           clearTimeout(maxMessageWaitTimer);
         }
-        // Resetting the mode. Now anyone can call start() or receive() again.
+
+        clearTimeout(waitTimer);
+
         if (this._receiver) {
           this._receiver.removeListener(ReceiverEvents.receiverError, onReceiveError);
           this._receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
           this._receiver.session.removeListener(SessionEvents.sessionError, onSessionError);
-        }
 
-        clearTimeout(waitTimer);
-        resetCreditWindow();
-        this.isReceivingMessages = false;
-        log.batching(
-          "[%s] Resolving the promise with received list of messages: %O.",
-          this._context.namespace.connectionId,
-          brokeredMessages
-        );
-        resolve(brokeredMessages);
+          if (this._receiver.credit > 0) {
+            log.batching("[%s] Draining leftover credits (%s).", this._context.namespace.connectionId, this._receiver.credit);
+
+            // Setting drain must be accompanied by a flow call (aliased to addCredit in this case).
+            this._receiver.drain = true;
+            this._receiver.addCredit(1);
```

Comment:
It does, but it will be drained off along with the rest of the credit. Rhea will not send a flow frame (addCredit is an alias for flow) if the value passed is 0, so we send 1. Relevant discussion: https://github.com/amqp/rhea/issues/156

Reply:
Thanks for the clarification! 

I would like to have @ramya-rao-a or @amarzavery's review on this as well -- they are currently on vacation. We'll try to get eyes on this early in the new year!